OpenRA.Game/Widgets/Widget.cs

```diff
@@ -289,25 +308,34 @@ public virtual bool TakeKeyboardFocus()
 			if (HasKeyboardFocus)
 				return true;
 
-			if (Ui.KeyboardFocusWidget != null && !Ui.KeyboardFocusWidget.YieldKeyboardFocus())
+			var k = Ui.GetCurrentKeyboardFocus();
+			if (k != null && !k.TryYieldKeyboardFocus())
 				return false;
 
-			Ui.KeyboardFocusWidget = this;
+			Ui.AddKeyboardFocus(this);
+			return true;
+		}
+
+		public virtual bool TryYieldKeyboardFocus()
+		{
 			return true;
 		}
 
 		public virtual bool YieldKeyboardFocus()
```

Comment:
Why are YieldKeyboardFocus() and TakeKeyboardFocus() virtual? Shouldn't only TryYieldKeyboardFocus() be virtual so a subclass can add conditions for yielding focus?